source/common/router/header_parser.cc

```diff
@@ -351,19 +358,34 @@ void HeaderParser::evaluateHeaders(Http::HeaderMap& headers,
       switch (entry.append_action_) {
         PANIC_ON_PROTO_ENUM_SENTINEL_VALUES;
       case HeaderValueOption::APPEND_IF_EXISTS_OR_ADD:
-        headers.addReferenceKey(key, value);
+        headers_to_add.emplace_back(key, value);
         break;
       case HeaderValueOption::ADD_IF_ABSENT:
         if (auto header_entry = headers.get(key); header_entry.empty()) {
-          headers.addReferenceKey(key, value);
+          headers_to_add.emplace_back(key, value);
         }
         break;
       case HeaderValueOption::OVERWRITE_IF_EXISTS_OR_ADD:
-        headers.setReferenceKey(key, value);
+        // TODO(cpakulski): Hash map is used to track multiple OVERWRITE actions
+        // and store only the last one. It would be more efficient to do it
+        // when config is processed. All OVERWRITE actions for the same header could be dropped
+        // except the last one.
+        headers_to_overwrite.erase(key);
+        headers_to_overwrite.try_emplace(key, key, value);
```

Comment:
Can `insert_or_assign` be used here?

Reply:
I tried it and it is not possible at the moment. It complains that `Http::LowerCaseString` has deleted copy constructor (or something around that). Let me investigate it more. I agree that `insert_or_assign` is cleaner than doing `erase` and `emplace`.

Reply:
i've never used try_emplace and do not understand why `key` is specified twice. But how about just writing `headers_to_overwrite[key] = value;` ?

Reply:
Refactored code not to use hash map, so this is not an issue anymore.